pkg/controller/certificaterequests/acme/acme.go

```diff
@@ -0,0 +1,213 @@
+	"k8s.io/client-go/tools/record"
+
+	"github.com/jetstack/cert-manager/pkg/acme"
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	cmclientset "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/certmanager/v1alpha1"
+	cmlisters "github.com/jetstack/cert-manager/pkg/client/listers/certmanager/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests"
+	crutil "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/util"
+	issuerpkg "github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CRControllerName = "certificaterequests-issuer-acme"
+)
+
+type ACME struct {
+	// used to record Events about resources to the API
+	recorder      record.EventRecorder
+	issuerOptions controllerpkg.IssuerOptions
+
+	orderLister cmlisters.OrderLister
+	cmClientV   cmclientset.CertmanagerV1alpha1Interface
+	helper      issuerpkg.Helper
+}
+
+func init() {
+	// create certificate request controller for acme issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		acme := NewACME(ctx)
+
+		orderInformer := ctx.SharedInformerFactory.Certmanager().V1alpha1().Orders().Informer()
+		controller := certificaterequests.New(apiutil.IssuerACME, acme, orderInformer)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewACME(ctx *controllerpkg.Context) *ACME {
+	ctx.CMClient.CertmanagerV1alpha1()
+	return &ACME{
+		recorder:      ctx.Recorder,
+		issuerOptions: ctx.IssuerOptions,
+		orderLister:   ctx.SharedInformerFactory.Certmanager().V1alpha1().Orders().Lister(),
+		cmClientV:     ctx.CMClient.CertmanagerV1alpha1(),
+	}
+}
+
+func (a *ACME) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuer v1alpha1.GenericIssuer) (*issuerpkg.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	reporter := crutil.NewReporter(cr, a.recorder)
+	resourceNamespace := a.issuerOptions.ResourceNamespace(issuer)
+
+	// If we can't decode the CSR PEM we have to hard fail
+	csr, err := pki.DecodeX509CertificateRequestBytes(cr.Spec.CSRPEM)
+	if err != nil {
+		message := "Failed to decode CSR in spec"
+
+		reporter.Failed(err, "ErrorParsingCSR", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	// If we fail to build the order we have to hard fail.
+	expectedOrder, err := buildOrder(cr, csr)
+	if err != nil {
+		message := "Failed to build order"
+
+		reporter.Failed(err, "ErrorBuildingOrder", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	order, err := a.orderLister.Orders(expectedOrder.Namespace).Get(expectedOrder.Name)
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
```

Comment:
Can you un-nest this, and then also `return nil, nil` at the end of the block? i.e.:

```go
if k8sErrors.IsNotFound(err) {
    ... existing code ...
    return nil, nil
}
if err != nil {
    ... existing code ...
}
```